File: sdk/ios/Sources/FraudShield/Collectors/BehavioralCollector.swift
import Foundation
import UIKit

struct BehavioralSignals {
    let sessionDurationMs: Double
    let keystrokeCount: Int
    let averageKeystrokeIntervalMs: Double
    let pasteDetected: Bool
    let pastedFields: [String]
    let recipientChangedCount: Int
    let transactionCreationMs: Double
    let typingSpeedScore: Double
}

class BehavioralCollector {

    private var sessionStartTime: Date?
    private var keystrokeTimes: [Date] = []
    private var pastedFields: [String] = []
    private var recipientChangedCount = 0

    func onSessionStart() {
        sessionStartTime = Date()
        keystrokeTimes.removeAll()
        pastedFields.removeAll()
        recipientChangedCount = 0

        // Observe UIPasteboard changes as fraud signal
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(pasteboardChanged),
            name: UIPasteboard.changedNotification,
            object: nil
        )
    }

    @objc private func pasteboardChanged() {
        pastedFields.append("pasteboard_change")
    }

    func recordKeystroke(field: String) {
        keystrokeTimes.append(Date())
    }

    func recordPaste(field: String) {
        pastedFields.append(field)
    }

    func recordRecipientChange() {
        recipientChangedCount += 1
    }

    func collect() async -> BehavioralSignals {
        let now = Date()
        let sessionDurationMs = sessionStartTime.map { now.timeIntervalSince($0) * 1000 } ?? 0

        // Calculate average time between keystrokes
        let intervals = zip(keystrokeTimes, keystrokeTimes.dropFirst())
            .map { $1.timeIntervalSince($0) * 1000 }
        let avgInterval = intervals.isEmpty ? 0.0 : intervals.reduce(0, +) / Double(intervals.count)

        let typingSpeedScore: Double
        switch avgInterval {
        case ..<50: typingSpeedScore = 0.1    // Extremely fast — scripted?
        case 50..<800: typingSpeedScore = 1.0 // Normal
        case 800..<3000: typingSpeedScore = 0.6 // Hesitant
        default: typingSpeedScore = avgInterval == 0 ? 0.5 : 0.3 // Very slow — coached?
        }

        return BehavioralSignals(
            sessionDurationMs: sessionDurationMs,
            keystrokeCount: keystrokeTimes.count,
            averageKeystrokeIntervalMs: avgInterval,
            pasteDetected: !pastedFields.isEmpty,
            pastedFields: pastedFields,
            recipientChangedCount: recipientChangedCount,
            transactionCreationMs: sessionDurationMs,
            typingSpeedScore: typingSpeedScore
        )
    }
}

File: sdk/android/src/main/kotlin/com/swifter/fraudshield/collectors/BehavioralCollector.kt
package com.swifter.fraudshield.collectors

data class BehavioralSignals(
    val sessionDurationMs: Long,
    val keystrokeCount: Int,
    val averageKeystrokeIntervalMs: Double,
    val pasteDetected: Boolean,
    val pastedFields: List<String>,
    val recipientChangedCount: Int,
    val transactionCreationMs: Long, // ms from session start to transaction submit
    val typingSpeedScore: Double    // 0.0 (very slow/nervous) to 1.0 (normal)
)

class BehavioralCollector {

    private var sessionStartMs: Long = 0L
    private val keystrokeTimes = mutableListOf<Long>()
    private val pastedFields = mutableListOf<String>()
    private var recipientChangedCount = 0

    fun onSessionStart() {
        sessionStartMs = System.currentTimeMillis()
        keystrokeTimes.clear()
        pastedFields.clear()
        recipientChangedCount = 0
    }

    fun recordKeystroke(fieldName: String) {
        keystrokeTimes.add(System.currentTimeMillis())
    }

    fun recordPaste(fieldName: String) {
        pastedFields.add(fieldName)
        keystrokeTimes.add(System.currentTimeMillis())
    }

    fun recordRecipientChange() {
        recipientChangedCount++
    }

    fun collect(): BehavioralSignals {
        val now = System.currentTimeMillis()
        val sessionDuration = if (sessionStartMs > 0) now - sessionStartMs else 0L
        val transactionCreationMs = sessionDuration

        // Calculate average interval between keystrokes
        val intervals = keystrokeTimes.zipWithNext { a, b -> (b - a).toDouble() }
        val avgInterval = if (intervals.isNotEmpty()) intervals.average() else 0.0

        // Typing speed score: very fast (<50ms avg) = suspicious (rushing),
        // very slow (>3000ms avg) = suspicious (nervous). Normal = 100-800ms
        val typingSpeedScore = when {
            avgInterval <= 0 -> 0.5 // No data
            avgInterval < 50 -> 0.1  // Extremely fast — scripted?
            avgInterval in 50.0..800.0 -> 1.0 // Normal
            avgInterval in 800.0..3000.0 -> 0.6 // Slow — hesitant?
            else -> 0.3 // Very slow — possibly coached
        }

        return BehavioralSignals(
            sessionDurationMs = sessionDuration,
            keystrokeCount = keystrokeTimes.size,
            averageKeystrokeIntervalMs = avgInterval,
            pasteDetected = pastedFields.isNotEmpty(),
            pastedFields = pastedFields.toList(),
            recipientChangedCount = recipientChangedCount,
            transactionCreationMs = transactionCreationMs,
            typingSpeedScore = typingSpeedScore
        )
    }
}
